service/src/main/java/org/cbioportal/service/impl/CoExpressionServiceImpl.java

```diff
@@ -28,71 +33,125 @@
     private MolecularDataService molecularDataService;
     @Autowired
     private GeneService geneService;
+    @Autowired
+    private GenesetService genesetService;
+    @Autowired
+    private GenesetDataService genesetDataService;
     
     @Override
-    public List<CoExpression> getCoExpressions(String molecularProfileId, String sampleListId, Integer entrezGeneId, 
-                                              Double threshold) throws MolecularProfileNotFoundException {
-
-        List<GeneMolecularData> molecularDataList = molecularDataService.getMolecularData(molecularProfileId, 
-            sampleListId, null, "SUMMARY");
+    public List<CoExpression> getCoExpressions(String geneticEntityId, String geneticEntityType, String sampleListId, 
+                                              String molecularProfileIdA, String molecularProfileIdB, Double threshold) throws Exception {
         
-        return createCoExpressions(molecularDataList, entrezGeneId, threshold);
+        List<CoExpression> computedCoExpressions = null;
+        List<? extends MolecularData>  molecularDataListA = null;
+        List<? extends MolecularData>  molecularDataListB = null;
+        if (geneticEntityType.toLowerCase().equals("gene")) {
+            molecularDataListA = molecularDataService.getMolecularData(molecularProfileIdA, sampleListId, null, "SUMMARY");
+        } else if (geneticEntityType.toLowerCase().equals("geneset")) {
+            molecularDataListA = genesetDataService.fetchGenesetData(molecularProfileIdA, sampleListId, null);
+        } else {
+                throw new Exception("Genetic Entity Type must be 'gene' or 'geneset'.");
+        }
+        if (molecularProfileIdB.matches("(.*)_gsva_scores")) {
```

Comment:
You should get the geneticAlterationType of this profile and check if it's equal to `MolecularAlterationType.GENESET_SCORE`, instead of checking the name of the profile.